aten/src/ATen/core/IList.h

```diff
@@ -0,0 +1,512 @@
+ * What is this for?
+ * =================
+ * Historically, PyTorch has maintained 2 different APIs: the unboxed
+ * (called from C++ API and Python eager mode) and boxed APIs (called
+ * from the TorchScript JIT, mobile interpreter, and boxed fallbacks).
+ *
+ * Calling unboxed kernels from the boxed "world" and vice-versa may
+ * result in non-negligible overhead. Lists are one of those types:
+ *
+ * - Boxed world: `c10::List`
+ * - Unboxed world: `c10::ArrayRef`
+ *
+ * In this context, `c10::IList` solves this problem by wrapping those
+ * 2 container types, so that we don't need to convert from one to
+ * the other.
+ *
+ * (see https://github.com/pytorch/pytorch/issues/66328)
+ *
+ * What does it do?
+ * ================
+ * This container wraps around the different tagged containers
+ * (currently, only boxed and unboxed), without incurring in extra
+ * overhead for converting from one to another. It does so while
+ * exposing usual container methods, which dispatch to corresponding
+ * implementations.
+ */
+template <typename T>
+class IList {
+ private:
+#define DEFINE_FRIEND_CLASS(TAG, ...)                  \
+  friend class detail::IListTagImpl<IListTag::TAG, T>; \
+  friend class detail::IListTagImplBase<               \
+      IListTag::TAG,                                   \
+      T,                                               \
+      typename detail::IListTagImpl<IListTag::TAG, T>::elem_type>;
+  TORCH_ILIST_FORALL_TAGS(DEFINE_FRIEND_CLASS)
+#undef DEFINE_FRIEND_CLASS
+
+ public:
+  using unboxed_type =
+      typename detail::IListTagImpl<IListTag::Unboxed, T>::list_type;
+  using boxed_type =
+      typename detail::IListTagImpl<IListTag::Boxed, T>::list_type;
+
+  using iterator = IListIterator<T>;
+  using const_iterator = IListIterator<T>;
+  using value_type = typename iterator::value_type;
+
+  IList() : tag_(IListTag::None) {}
+
+  IList(const boxed_type& boxed) : tag_(IListTag::Boxed) {
+    payload_.boxed = &boxed;
+  }
+
+  IList(const unboxed_type& unboxed) : tag_(IListTag::Unboxed) {
+    payload_.unboxed = unboxed;
+  }
+
+  /*
+   * Enable only 2 implicit constructors:
+   * 1. IList(const std::vector<T>&)
+   * 2. IList(std::initializer_list<T>)
+   *
+   * These are only enabled if the `unboxed_type` also has a constructor
+   * for those types.
+   */
+  IList(
+      const std::vector<T>& vec,
+      std::enable_if_t<
+          std::is_constructible<unboxed_type, std::vector<T>>::value,
+          std::nullptr_t> = nullptr)
+      : tag_(IListTag::Unboxed) {
+    payload_.unboxed = unboxed_type(vec);
+  }
+
+  constexpr IList(
+      std::initializer_list<T> il,
+      std::enable_if_t<
+          std::is_constructible<unboxed_type, std::initializer_list<T>>::value,
+          std::nullptr_t> = nullptr)
+      : tag_(IListTag::Unboxed) {
+    payload_.unboxed = unboxed_type(il);
+  }
+
+  /*
+   * For convenience, enable other supported `unboxed_type` constructors as
+   * explicit constructors.
+   */
```

Comment:
Hmm, I'm confused. Don't you need to support *all* implicit constructors from ArrayRef for IList to be a drop-in replacement for ArrayRef?

Reply:
Yeah. This came before we discussed in the other PR. I will make them implicit.